### PR TITLE
Add function for checking QoS profile compatibility

### PR DESCRIPTION
**Pull request by jacobperron**
Given QoS profiles for a publisher and a subscription, this function
will return `true` if the profiles are compatible.
Compatible profiles implies that the subscription and publisher will
be able to communicate.

Connected to https://github.com/ros2/rmw/pull/299

### CODE REVIEW

**Comment by jacobperron** [comment]
This conflicts with https://github.com/ros2/rmw_dds_common/pull/43, so I'll likely have to rebase once that lands.

**Comment by jacobperron** [comment]
Compatibility checks are based on the tables found here: https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/#qos-compatibilities

File: rmw_dds_common/src/qos.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <sstream>
+
+#include "rmw/qos_profiles.h"
+
+#include "rmw_dds_common/qos.hpp"
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:20]:
```suggestion
#include "rmw_dds_common/qos.hpp"

#include <string>
#include <sstream>

#include "rmw/qos_profiles.h"
```

The headers order is: related header, system headers, other headers.

File: rmw_dds_common/src/qos.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+static std::ostream &
+operator<<(std::ostream & out, const rmw_time_t & t)
+{
+  out << "{" << t.sec << "s " << t.nsec << "ns}";
+  return out;
+}
+
+static void
+_write_to_buffer(const std::string message, char * buffer, size_t buffer_size)
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:56]:
```suggestion
_write_to_buffer(const char * message, char * buffer, size_t buffer_size)
```

the first argument can be a cstring.

L237 needs to change to:   `_write_to_buffer(reason_ss.str().c_str(), reason, reason_size);`

File: rmw_dds_common/src/qos.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+rmw_ret_t
+qos_profile_check_compatible(
+  const rmw_qos_profile_t publisher_qos,
+  const rmw_qos_profile_t subscription_qos,
+  rmw_qos_compatibility_type_t * compatibility,
+  char * reason,
+  size_t reason_size)
+{
+  if (!compatibility) {
+    _write_to_buffer("ERROR: compatibility parameter is null", reason, reason_size);
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:79]: I would expect reason to only be set if the return value is `RMW_RET_OK` and the profiles are incompatible.
For error messages use `RMW_SET_ERROR_MSG()`.

e.g. the functions for validation a name/namespace/topic name work in a similar fashion.

File: rmw_dds_common/src/qos.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  std::ostringstream reason_ss;
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:83]: I see that the point of having a `reason` and `reason_size` arguments is to avoid allocations, but a `std::ostringstream` will actually allocate in the heap.

File: rmw_dds_common/src/qos.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  strncpy(buffer, message.c_str(), write_size);
+}
+
+rmw_ret_t
+qos_profile_check_compatible(
+  const rmw_qos_profile_t publisher_qos,
+  const rmw_qos_profile_t subscription_qos,
+  rmw_qos_compatibility_type_t * compatibility,
+  char * reason,
+  size_t reason_size)
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:76]: I would expect `reason=nullptr, reason_size=3` to be reported as an error.

File: rmw_dds_common/src/qos.cpp (hunk 4 above)
Comment (ivanpauno) [reply rmw_dds_common/src/qos.cpp:83]: you can instead use `rcutils_snprintf` like [here](https://github.com/ros2/rcutils/blob/1eae6891b2ef54563ab6d11d564ebdabc7939398/include/rcutils/error_handling.h#L235-L244).

File: rmw_dds_common/src/qos.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+  }
+
+  std::ostringstream reason_ss;
+
+  // Presume profiles are compatible until proven otherwise
+  *compatibility = RMW_QOS_COMPATIBILITY_OK;
+
+  // If there are any "unknown" values, then there is an error
+  if (RMW_QOS_POLICY_RELIABILITY_UNKNOWN == publisher_qos.reliability) {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:90]: this seems to be an invalid argument and not a compatibility error

File: rmw_dds_common/src/qos.cpp [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    reason_ss << "ERROR: Publisher's liveliness is automatic and subscription's liveliness is " <<
+      "manual by topic;";
+  }
+
+  const rmw_time_t & pub_lease = publisher_qos.liveliness_lease_duration;
+  const rmw_time_t & sub_lease = subscription_qos.liveliness_lease_duration;
+  const rmw_time_t lease_default = RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT;
+
+  // No lease duration for publisher and lease duration for subscription
+  if (pub_lease == lease_default && sub_lease != lease_default) {
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:169]: I'm not so sure of this.
IIUC, the default is the [participant liveliness lease duration](https://community.rti.com/static/documentation/connext-dds/6.0.1/doc/api/connext_dds/api_cpp/structDDS__DiscoveryConfigQosPolicy.html#a78c4d3b8bda361a00e123f58ea496e4c), so the profiles could still be compatible (with a suitable participant liveliness lease duration).

**Review by ivanpauno — CHANGES_REQUESTED** [review]

File: rmw_dds_common/src/qos.cpp (hunk 4 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:83]: Good catch, I missed that!

File: rmw_dds_common/src/qos.cpp (hunk 1 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: Fixed in cc81b3f

File: rmw_dds_common/src/qos.cpp (hunk 2 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: Updated in 132b3e1

File: rmw_dds_common/src/qos.cpp (hunk 3 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: Done in 1a156da

File: rmw_dds_common/src/qos.cpp (hunk 4 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: I've refactored to avoid using ostringstream in 9c43a6e

File: rmw_dds_common/src/qos.cpp (hunk 5 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:75]: Handled this case in e7d4b0a

File: rmw_dds_common/src/qos.cpp (hunk 6 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: No longer setting compatibility result in error cases: 70dc190

File: rmw_dds_common/src/qos.cpp (hunk 7 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:193]: The RMW docs claim that this default value implies liveliness is not tracked at all: https://github.com/ros2/rmw/blob/ae0553adb574c031343965195ec4a73039e62389/rmw/include/rmw/types.h#L433-L434

I haven't confirmed if that is true though.

Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:193]: For rmw_cyclonedds, the value of zero (the default) results in an ["infinite" duration](https://github.com/ros2/rmw_cyclonedds/blob/d024823043504ea40af24bf22365a21cd203df55/rmw_cyclonedds_cpp/src/rmw_node.cpp#L1709-L1712).

For rmw_fastrtps and rmw_connext, we seem to be defaulting to their defaults, only setting a value if it is not the rmw default:

https://github.com/ros2/rmw_fastrtps/blob/901b0ed7d286b34cda72b3a1138c6e05bf1096c4/rmw_fastrtps_shared_cpp/src/qos.cpp#L125

https://github.com/ros2/rmw_connext/blob/fd3e6a71fee0f7eb26f84505d16169ce489d6235/rmw_connext_shared_cpp/src/qos.cpp#L119

Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:193]: I took a look at the values default values for the talker demo with `ros2 topic info --verbose /chatter`, and the RMWs report very large default values (which I think effectively means liveliness is not tracked):

Value for Fast-DDS and CycloneDDS: 9223372036854775807 nanoseconds
Value for Connext: 2147483649147483647 nanoseconds

**Comment by jacobperron** [comment]
@ivanpauno Thanks for the review. I think I've address all of your comments.

**Review by audrow — APPROVED** [review]
This looks good to me. The QoS compatability check function makes sense, and the tests do a nice job to confirm that results are as expected for different QoS profile combinations. Nice job!

File: rmw_dds_common/src/qos.cpp (hunk 7 above)
Comment (ivanpauno) [reply rmw_dds_common/src/qos.cpp:193]: Liveliness is always somehow tracked, but it seems that the matching rules are fine.

**Review by ivanpauno — APPROVED** [review]

File: rmw_dds_common/include/rmw_dds_common/qos.hpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * This implements the rmw API \ref rmw_qos_profile_check_compatible().
+ * See \ref rmw_qos_profile_check_compatible() for more information.
+ *
+ * \param[in] publisher_profile: The QoS profile used for a publisher.
+ * \param[in] subscription_profile: The QoS profile used for a subscription.
+ * \param[out] compatibility: `RMW_QOS_COMPATIBILITY_OK` if the QoS profiles are compatible, or
+ *   `RMW_QOS_COMPATIBILITY_WARNING` if the QoS profiles might be compatible, or
+ *   `RMW_QOS_COMPATIBILITY_ERROR` if the QoS profiles are not compatible.
+ * \param[out] reason: A detailed reason for a QoS incompatibility.
+ *   Must be pre-allocated by the caller. This parameter is optional and may be set to `NULL`.
```
Comment (mjeronimo) [inline rmw_dds_common/include/rmw_dds_common/qos.hpp:40]: Should this be "nullptr" instead of "NULL"?

Also, perhaps: "This parameter is optional and my be set to 'nullptr' if 'reason' information is not desired."

File: rmw_dds_common/include/rmw_dds_common/qos.hpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ * using the QoS policies can communicate with each other.
+ *
+ * This implements the rmw API \ref rmw_qos_profile_check_compatible().
+ * See \ref rmw_qos_profile_check_compatible() for more information.
+ *
+ * \param[in] publisher_profile: The QoS profile used for a publisher.
+ * \param[in] subscription_profile: The QoS profile used for a subscription.
+ * \param[out] compatibility: `RMW_QOS_COMPATIBILITY_OK` if the QoS profiles are compatible, or
+ *   `RMW_QOS_COMPATIBILITY_WARNING` if the QoS profiles might be compatible, or
+ *   `RMW_QOS_COMPATIBILITY_ERROR` if the QoS profiles are not compatible.
```
Comment (mjeronimo) [inline rmw_dds_common/include/rmw_dds_common/qos.hpp:38]: Re-reading the description, it's not quite clear to me if reason information is available in the "may-or-may-not-be-compatible" case. Perhaps: "A detailed reason for a QoS incompatibility or potential incompatibility."

File: rmw_dds_common/include/rmw_dds_common/qos.hpp [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+ * \param[out] reason: A detailed reason for a QoS incompatibility.
```
Comment (mjeronimo) [inline rmw_dds_common/include/rmw_dds_common/qos.hpp:39]: I noticed that the test code is initializing the 'reason' buffer before calling this function:

`    char reason[2048] = "";`


If this is a requirement for client code, it could be mentioned here. Otherwise, the implementation could also initialize the buffer at the start in case the user declares the buffer on the stack without initialization:

`    char reason[2048];`

and the function appends the message to some junk on the stack.

File: rmw_dds_common/include/rmw_dds_common/qos.hpp [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+#include "rmw/types.h"
+
+#include "rmw_dds_common/visibility_control.h"
+
+namespace rmw_dds_common
+{
+
+/// Check if two QoS profiles are compatible
+/**
+ * Two QoS profiles are compatible if a publisher and subcription
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp:28]: TBH, I can't recall why that terminology. I'm just following suite with the terminology used throughout the rest of the ROS 2 stack, though I think we're also inconsistent about what terminology we're using :upside_down_face: The tutorials seem to use "subscriber" instead of "subscription".

[the remark this answers is not in the document]
Comment (jacobperron) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp:28]: @wjwwood Probably has some insight.

File: rmw_dds_common/include/rmw_dds_common/qos.hpp (hunk 9 above)
Comment (jacobperron) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp:38]: I purposefully left some info out to avoid too much redundancy with the rmw docs (instead just referencing them). I think the docs in https://github.com/ros2/rmw/pull/299/ capture clarify this, but I can copy over the details here if you think it's better.

File: rmw_dds_common/include/rmw_dds_common/qos.hpp (hunk 10 above)
Comment (jacobperron) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp:39]: Good catch! I think doing the zero-initialization of the buffer in the function makes sense; one less thing for the user to worry about.

File: rmw_dds_common/include/rmw_dds_common/qos.hpp (hunk 9 above)
Comment (jacobperron) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp:38]: I've incorporated your suggestion: 26f2e63

File: rmw_dds_common/include/rmw_dds_common/qos.hpp (hunk 8 above)
Comment (jacobperron) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp]: Addressed: 26f2e63

File: rmw_dds_common/include/rmw_dds_common/qos.hpp (hunk 10 above)
Comment (jacobperron) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp]: See 530c42f

File: rmw_dds_common/src/qos.cpp [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+#include <cstring>
+
+#include "rcutils/snprintf.h"
+#include "rmw/error_handling.h"
+#include "rmw/qos_profiles.h"
+
+namespace rmw_dds_common
+{
+
+static bool
```
Comment (mjeronimo) [inline rmw_dds_common/src/qos.cpp:26]: IMO, these rmw_time-related operators could live in a publicly-available header as potentially useful functionality outside of this QoS comparison functionality.

Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:26]: I agree. But, there is some on-going discussion about deprecating `rmw_time_t` (https://github.com/ros2/rmw/pull/298) that would require changes here. I'd rather wait to see how that plays out before abstracting these operators to a common place.

File: rmw_dds_common/src/qos.cpp [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+      return append_ret;
+    }
+  }
+
+  const rmw_time_t & pub_deadline = publisher_qos.deadline;
+  const rmw_time_t & sub_deadline = subscription_qos.deadline;
+  const rmw_time_t deadline_default = RMW_QOS_DEADLINE_DEFAULT;
+
+  // No deadline for publisher and deadline for subscription
+  if (pub_deadline == deadline_default && sub_deadline != deadline_default) {
```
Comment (eboasson) [inline rmw_dds_common/src/qos.cpp:154]: The policies in ROS have this annoying `SYSTEM_DEFAULT` thing which I have always read as the application explicitly "don't care, do what you think is best", which doesn't make much sense to me. Secondly, I've always interpreted that setting as valid only when creating an endpoint, but never as a value returned by the underlying middleware.

If that interpretation is correct (which it may not be, given that the `SYSTEM_DEFAULT` is handled later in this function), then it seems illogical that a default deadline setting would be treated differently here. Doesn't it make far more sense to simply say the default deadline is ∞ (however represented)? That way, there always is a deadline.

If you then define the comparison functions such that ∞ is handled correctly, this all becomes much simpler, IMHO.

**Review by eboasson — APPROVED** [review]

File: rmw_dds_common/src/qos.cpp (hunk 13 above)
Comment (ivanpauno) [reply rmw_dds_common/src/qos.cpp:154]: > Secondly, I've always interpreted that setting as valid only when creating an endpoint, but never as a value returned by the underlying middleware.

Yes, that should be the case.

> If that interpretation is correct (which it may not be, given that the SYSTEM_DEFAULT is handled later in this function), then it seems illogical that a default deadline setting would be treated differently here. Doesn't it make far more sense to simply say the default deadline is ∞ (however represented)? That way, there always is a deadline.

But the "system default" could be different from infinite, e.g. if you specify in a qos profile file a default profile with a different deadline value (which would be weird, but possible).
I would only use this function with "actual" qos profiles , and not with one where an item can be "SYSTEM_DEFAULT", so we can also consider failing if an item is "SYSTEM_DEFAULT".

Comment (ivanpauno) [reply rmw_dds_common/src/qos.cpp:154]: Actually, this new function is failing when a policy has value "UNKNOWN", which is used when an entity has a qos value valid for the underlying middleware that hasn't an equivalent in ROS 2.

I'm not sure why this function should return a compatibility warning when "SYSTEM_DEFAULT" and an error when "UNKNOWN".
I actually think that the opposite makes more sense, or either returning a compatibility warning or error in both.

@jacobperron what do you think?

File: rmw_dds_common/src/qos.cpp [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+      rmw_ret_t ret = _append_to_buffer(
+        "WARNING: Publisher's liveliness is automatic, but subscription's is system default;",
+        reason,
+        reason_size);
+      if (RMW_RET_OK != ret) {
+        return ret;
+      }
+    } else {
+      if (publisher_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT &&
+        subscription_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC)
```
Comment (jacobperron) [inline rmw_dds_common/src/qos.cpp:295]: I just realized there's another case where we don't know if the profiles are valid, if both the publisher and subscription are "system default".

File: rmw_dds_common/src/qos.cpp (hunk 13 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:154]: > But the "system default" could be different from infinite, e.g. if you specify in a qos profile file a default profile with a different deadline value (which would be weird, but possible).

I wasn't sure if `RMW_QOS_DEADLINE_DEFAULT` is equivalent to "system default". If it is, then I agree we should treat it the same as other system default values (e.g. warn).

> I would only use this function with "actual" qos profiles , and not with one where an item can be "SYSTEM_DEFAULT", so we can also consider failing if an item is "SYSTEM_DEFAULT".

Unfortunately, all of the [widely used profiles defined in `rmw`](https://github.com/ros2/rmw/blob/ae0553adb574c031343965195ec4a73039e62389/rmw/include/rmw/qos_profiles.h) use "SYSTEM_DEFAULT" values. I think it would be unfortunate if this function failed doing something like this:

```c
rmw_qos_profile_check_compatible( rmw_qos_profile_sensor_data, rmw_qos_profile_sensor_data, ...);
```

Although, producing a warning isn't much better...

Maybe we should change the default values for those profiles :sweat_smile: 

> Actually, this new function is failing when a policy has value "UNKNOWN", which is used when an entity has a qos value valid for the underlying middleware that hasn't an equivalent in ROS 2.

I don't know if this applies here. This is a common DDS implementation that presumes the RMW supports all the ROS 2 QoS settings (we could document it here to be explicit). If an RMW doesn't support all of the QoS policies, then they should probably provide their own implementation of this function, especially if the RMW is not a DDS implementation.

File: rmw_dds_common/src/qos.cpp (hunk 14 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:295]: Maybe I'm overthinking this one. If both are system default, then presumably they are using the same underlying QoS settings, though I guess it's possible the policy for a subscription can still be configured independently from the publisher s.t. they are incompatible (that would be unfortunate).

Comment (ivanpauno) [reply rmw_dds_common/src/qos.cpp:295]: > then presumably they are using the same underlying QoS settings

The system default is read from a qos profile file (if there is one), if not it's the DDS defaults.
You could potentially write a qos profile file where the default profile for publishers is incompatible with the one for subscriptions, but that would be bizarre ...

File: rmw_dds_common/src/qos.cpp (hunk 13 above)
Comment (ivanpauno) [reply rmw_dds_common/src/qos.cpp:154]: What is this function for?

I imagine it being used in two ways:

- Check if the qos profile of two existing endpoints is compatible (for example, using the qos profiles got from the `rmw_get_publishers_info_by_topic`/`rmw_get_subscription_info_by_topic`).
  This use case doesn't seem super useful though, better to use the "Qos incompatible event".
- Check if an user provided profile is compatible with the profile of an existing(s) endpoint (one profile got from the functions above, the other is user provided).
  Useful if you want to "make up" a profile compatible with existing entities.

In case 1, "SYSTEM_DEFAULT" is impossible, and "UNKNOWN" is a possible value.
In case 2, the user provided profile can have a "SYSTEM_DEFAULT" value, but I'm not sure why someone would pass that.

> I don't know if this applies here. This is a common implementation for DDS that presumes the RMW supports all the ROS 2 QoS settings (we could document it here to be explicit). If an RMW doesn't support all of the QoS policies

Yes, and in that case you can still get "UNKNOWN".
e.g.: 
- create a publisher with durability=SYSTEM_DEFAULT
- pass a qos profile file that specifies "DDS_TRANSIENT_DURABILITY_QOS" (using topic filters if desired) (transient is not the same as transient local)
- call `rmw_get_publishers_info_by_topic()`, the qos profile returned will have `durability=UNKNOWN`.

Same applies to any application mixing native DDS with ROS 2.

Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:154]: >  Doesn't it make far more sense to simply say the default deadline is ∞ (however represented)? That way, there always is a deadline.

@eboasson I think this is what https://github.com/ros2/rmw/pull/301 is trying to address. Moving away from a "default" value to a more explicit "infinite" value.

AFAICT, passing a "default" value for deadline, liveliness, or lifespan will result in an "infinite" duration (for cyclonedds, fast-dds, and connext). However, this is not what is [documented](https://github.com/ros2/rmw/blob/ae0553adb574c031343965195ec4a73039e62389/rmw/include/rmw/types.h#L427) in `rmw`, hence this special logic. If #301 lands, then I'm inclined to change this implementation as you pointed out.

Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:154]: @ivanpauno Yeah, those are the two uses of this function that I had in mind. And the first one is the primary motivation for this API. AFAIK, tools that want to introspect the ROS graph (e.g. ros2doctor and rqt_graph) cannot get "QoS incompatible" events for all nodes in the graph. This API let's them warn about (possible) compatibility issues.

I guess it is technically possible to get a "SYSTEM_DEFAULT" back from `rmw_get_publishers_info_by_topic`. Though it's not very useful, there's nothing that says it can't return such a value. My thought is that this function will report possible compatibility issues to the best of it's ability. If it gets a policy value of "SYSTEM_DEFAULT" or "UNKNOWN" then the best we can do is warn about possible incompatibilities. Originally, I didn't think that it was possible to get "UNKNOWN" values from `rmw_get_publishers_info_by_topic` / `rmw_get_subscription_info_by_topic`, but if that is the case then I agree we should change the behavior of this function to produce a warning instead.

File: rmw_dds_common/src/qos.cpp (hunk 14 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:295]: > You could potentially write a qos profile file where the default profile for publishers is incompatible with the one for subscriptions, but that would be bizarre ...

Yeah, I agree; bizarre. But, given this information we can't do any better than say there is a possible compatibility issue, so I think the best thing to do is warn in this case.

File: rmw_dds_common/src/qos.cpp (hunk 13 above)
Comment (ivanpauno) [reply rmw_dds_common/src/qos.cpp:154]: > AFAIK, tools that want to introspect the ROS graph (e.g. ros2doctor and rqt_graph) cannot get "QoS incompatible" events for all nodes in the graph

Ah yes, that's true.

> I guess it is technically possible to get a "SYSTEM_DEFAULT" back from rmw_get_publishers_info_by_topic

`rmw_get_publishers_info_by_topic` should not return "SYSTEM_DEFAULT", but a more specific value.

> but if that is the case then I agree we should change the behavior of this function to produce a warning instead.

Using a warning for both system default and unknown sounds good.

File: rmw_dds_common/src/qos.cpp (hunk 12 above)
Comment (mjeronimo) [reply rmw_dds_common/src/qos.cpp:27]: Ok, sounds good.

**Review by mjeronimo — APPROVED** [review]

File: rmw_dds_common/src/qos.cpp (hunk 13 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:132]: Warning on "unknown" values: 6c789b4

File: rmw_dds_common/src/qos.cpp [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+    // Reliability for publisher is unknown and subscription is reliable
+    if (pub_reliability_unknown &&
+      subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE)
+    {
+      *compatibility = RMW_QOS_COMPATIBILITY_WARNING;
+      rmw_ret_t append_ret = _append_to_buffer(
+        reason,
+        reason_size,
+        "WARNING: Reliable subscription, but publisher is %s;",
+        rmw_qos_reliability_policy_to_str(publisher_qos.reliability));
```
Comment (jacobperron) [inline rmw_dds_common/src/qos.cpp:232]: I was writing some more  tests and discovered the "*to_str" functions return `NULL` if the policy is unknown... I need to refactor this.

File: rmw_dds_common/include/rmw_dds_common/qos.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (wjwwood) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp:28]: A publisher publishes, a subscription is a receipt of a subscription, it isn't a subscriber. So subscriber is the wrong term. A node is kind of a subscribe as it is the thing that subscribes. A publication could be a thing, but because our object publishes, it is a publisher.

File: rmw_dds_common/src/qos.cpp (hunk 13 above)
Comment (wjwwood) [reply rmw_dds_common/src/qos.cpp:133]: > > I guess it is technically possible to get a "SYSTEM_DEFAULT" back from rmw_get_publishers_info_by_topic

> rmw_get_publishers_info_by_topic should not return "SYSTEM_DEFAULT", but a more specific value.

I think that `rmw_get_publishers_info_by_topic` should try and avoid returning system default, or shouldn't ever. The idea is that system default is a stand-in for "some externally specified setting", so when querying it for a real entity, it should just return the ROS equivalent or unknown if there isn't an equivalent.

File: rmw_dds_common/src/qos.cpp (hunk 15 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: Fixed in 6a67320

File: rmw_dds_common/src/qos.cpp (hunk 14 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: Handled the case where we have two "system default" / "unknown" values: 6a67320

File: rmw_dds_common/include/rmw_dds_common/qos.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (mjeronimo) [reply rmw_dds_common/include/rmw_dds_common/qos.hpp:28]: @wjwwood  Interesting. I guess I just think of it more like this (probably attracted to the symmetry/uniformity):



Anyways, thanks for the info. It helps to fill in some of the blanks.

File: rmw_dds_common/src/qos.cpp [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+  if (!reason && reason_size != 0u) {
+    RMW_SET_ERROR_MSG("reason parameter is null, but reason_size parameter is not zero");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  // Presume profiles are compatible until proven otherwise
+  *compatibility = RMW_QOS_COMPATIBILITY_OK;
+
+  // Initialize reason buffer
+  if (reason) {
```
Comment (fujitatomoya) [inline rmw_dds_common/src/qos.cpp:96]:
```suggestion
  if (reason && reason_size != 0u) {
```

i think there is a possibility that user can input `reason is not NULL` and `reason_size = 0`. (unlikely but possible) in that case, this overruns the reason_size. i think it is expected to do nothing if reason_size is 0.

or we could return RMW_RET_INVALID_ARGUMENT in this case, since user arguments do not make sense.

File: rmw_dds_common/src/qos.cpp [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+  }
+
+  // Best effort publisher and reliable subscription
+  if (publisher_qos.reliability == RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT &&
+    subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE)
+  {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    rmw_ret_t append_ret = _append_to_buffer(
+      reason,
+      reason_size,
```
Comment (fujitatomoya) [inline rmw_dds_common/src/qos.cpp:107]: isn't this supposed to be `(reason_size - 1)`? the null terminator will be written outside the end of the array, overwriting memory not belonging to the array.

Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:107]: According the the [docs of snprintf](https://en.cppreference.com/w/c/io/fprintf), it will naturally account for the null character:

> At most bufsz - 1 characters are written

So I don't think we have to handle that here.

File: rmw_dds_common/src/qos.cpp (hunk 16 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp]: Good catch.

We should definitely guard against zero size here. I think snprintf handles size `0` correctly (no bytes are written). I'm inclined to keep an "OK" return code, since a lot of standard library functions seem to be okay with zero lengths.

Applied your suggestion with a unit test in a8862ee

File: rmw_dds_common/src/qos.cpp (hunk 17 above)
Comment (fujitatomoya) [reply rmw_dds_common/src/qos.cpp:107]: i think that is correct, thanks :+1:

**Review by fujitatomoya — APPROVED** [review]

File: rmw_dds_common/src/qos.cpp (hunk 13 above)
Comment (jacobperron) [reply rmw_dds_common/src/qos.cpp:133]: To circle back to the original comment, I think I'll leave the current implementation as-is since it seems to agree with the RMW API docs and https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/#qos-compatibilities.
However, I do think that replacing the default duration value with an explicit infinity value would be better (related to https://github.com/ros2/rmw/pull/301). Pending further changes to the RMW API, I'm happy to come back and simplify the logic in this implementation.

While endpoints may not return "SYSTEM_DEFAULT" values, it's possible that other use-cases not involving endpoint queries will call this function. So I think it's nice to warn about "SYSTEM_DEFAULT"  values, instead of returning an error with no info about other possible incompatibilities.

**Comment by jacobperron** [comment]
CI: https://github.com/ros2/rmw/pull/299#issue-571521564

**Comment by jacobperron** [comment]
@ros-pull-request-builder retest this please

**Comment by jacobperron** [comment]
@ros-pull-request-builder retest this please
